client/reader/stream/post-lifecycle.jsx

```diff
@@ -17,6 +18,63 @@ import PostUnavailable from './post-unavailable';
 import RecommendedPosts from './recommended-posts';
 import CrossPost from './x-post';
 
+/**
+ * Hook to return a [callback ref](https://reactjs.org/docs/refs-and-the-dom.html#callback-refs)
+ * that MUST be used as the `ref` prop on a `div` element.
+ * The hook ensures that we generate post display Tracks events when the user views
+ * the underlying `div` element.
+ * @param postObj Object The post data.
+ * @returns A callback ref that MUST be used on a div element for tracking.
+ */
+const useTrackPostView = ( postObj ) => {
+	const observerRef = useRef();
+
+	// Use a callback as the ref so we get called for both mount and unmount events
+	// We don't get both if we use useRef() and useEffect() together.
+	return useCallback(
+		( wrapperDiv ) => {
+			// If we don't have a wrapper div, we aren't mounted and should remove the observer
+			if ( ! wrapperDiv ) {
+				observerRef.current?.disconnect?.();
+				return;
+			}
+
+			const intersectionHandler = ( entries ) => {
+				const [ entry ] = entries;
+				if ( ! entry.isIntersecting ) {
+					return;
+				}
+
+				recordTracksEvent( 'calypso_reader_post_display', {
+					feed_id: postObj?.feed_ID,
+					site_id: postObj?.blogId,
+					post_id: postObj?.postId,
```

Comment:
Might we be able to include information about which page/context we're showing the post in? That would be super-useful to include. (I believe we track that in the `ui_algo` event prop for other Reader Tracks events.)

Reply:
Sure! It looks like we have access to the `streamKey` property which tells us which stream/tags are being shown; I've added it as the `ui_algo` prop.

Reply:
I am not sure if that is the correct `ui_algo` value that we need to be sending. 
Maybe here it's a good place to use [this method](https://github.com/Automattic/wp-calypso/blob/trunk/client/reader/stats.js#L113)?

Reply:
Ooh that's a nice find! Let's see if I can refactor to use it instead of setting all the props manually.

Reply:
This looks better, I will re-test!